parallel-20180522/src/env_parallel.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.32)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "ENV_PARALLEL 1"
.TH ENV_PARALLEL 1 "2018-05-21" "20180422" "parallel"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
env_parallel \- export environment to GNU parallel
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBenv_parallel\fR [options for \s-1GNU\s0 Parallel]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBenv_parallel\fR is a shell function that exports the current
environment to \s-1GNU \s0\fBparallel\fR.
.PP
If the shell function is not loaded, a dummy script will be run
instead that explains how to install the function.
.PP
\&\fBenv_parallel\fR is beta quality and not production ready, but please
use it for everyday use and report bugs.
.PP
\&\fBenv_parallel\fR is 100 ms slower at startup than pure \s-1GNU
\&\s0\fBparallel\fR, and takes up to 30% longer to start a job (typically 15 ms).
.PP
Due to the problem with environment space (see below) the recommended
usage is either:
.PP
.Vb 2
\&  # Do \-\-record\-env into $PARALLEL_IGNORED_NAMES
\&  env_parallel \-\-session
\&
\&  # Define whatever you want to use
\&  alias myalias=echo
\&  myvar=it
\&  myfunc() { myalias $1 $myvar works.; }
\&
\&  # env_parallel will not export names in $PARALLEL_IGNORED_NAMES
\&  env_parallel \-S localhost myfunc ::: Yay,
.Ve
.PP
Or:
.PP
.Vb 2
\&  # Record the "clean" environment (this only needs to be run once)
\&  env_parallel \-\-record\-env
\&
\&  # Optionally edit ~/.parallel/ignored_vars (only needed once)
\&
\&  # Define whatever you want to use
\&  alias myalias=echo
\&  myvar=it
\&  myfunc() { myalias $1 $myvar works.; }
\&
\&  # Use \-\-env _ to only transfer the names not in the "empty" environment
\&  env_parallel \-\-env _ \-S localhost myfunc ::: Yay,
.Ve
.PP
In \fBcsh\fR \fB\-\-session\fR is not supported:
.PP
.Vb 2
\&  # Record the "clean" environment (this only needs to be run once)
\&  env_parallel \-\-record\-env
\&
\&  # Optionally edit ~/.parallel/ignored_vars (only needed once)
\&
\&  # Define whatever you want to use
\&  alias myalias \*(Aqecho \e!* $myvar works.\*(Aq
\&  set myvar=it
\&
\&  # Use \-\-env _ to only transfer the names not in the "empty" environment
\&  env_parallel \-\-env _ \-S localhost myalias ::: Yay,
.Ve
.SS "Environment space"
.IX Subsection "Environment space"
By default \fBenv_parallel\fR will export all environment variables,
arrays, aliases, functions and shell options (see details for the
individual shells below).
.PP
But this only works if the size of the current environment is smaller
than the maximal length of a command and smaller than half of the max
if running remotely. E.g. The max size of Bash's command is 128 \s-1KB,\s0 so
\&\fBenv_parallel\fR will fail if '\fBset | wc \-c\fR' is bigger than 128
\&\s-1KB.\s0 Technically the limit is in \fIexecve\fR\|(1) which IPC::open3 uses.
.PP
Bash completion functions are well-known for taking up well over 128
\&\s-1KB\s0 of environment space and the primary reason for causing
\&\fBenv_parallel\fR to fail.
.PP
Instead you can use \fB\-\-env\fR to specify which variables, arrays,
aliases and functions to export as this will only export those with
the given name. Or follow the recommended usage in shown in
\&\s-1DESCRIPTION.\s0
.SH "OPTIONS"
.IX Header "OPTIONS"
Same as \s-1GNU \s0\fBparallel\fR.
.SH "SUPPORTED SHELLS"
.IX Header "SUPPORTED SHELLS"
.SS "Ash"
.IX Subsection "Ash"
\fIInstallation\fR
.IX Subsection "Installation"
.PP
Put this in \f(CW$HOME\fR/.profile:
.PP
.Vb 1
\&  . \`which env_parallel.ash\`
.Ve
.PP
E.g. by doing:
.PP
.Vb 1
\&  echo \*(Aq. \`which env_parallel.ash\`\*(Aq >> $HOME/.profile
.Ve
.PP
\fISupported use\fR
.IX Subsection "Supported use"
.PP
\&\fB\-\-env\fR is supported to export only the variable, or alias with the
given name. Multiple \fB\-\-env\fRs can be given.
.IP "aliases" 8
.IX Item "aliases"
.Vb 5
\&  alias myecho=\*(Aqecho aliases\*(Aq
\&  env_parallel myecho ::: work
\&  env_parallel \-S server myecho ::: work
\&  env_parallel \-\-env myecho myecho ::: work
\&  env_parallel \-\-env myecho \-S server myecho ::: work
\&
\&  alias multiline=\*(Aqecho multiline
\&    echo aliases\*(Aq
\&  env_parallel multiline ::: work
\&  env_parallel \-S server multiline ::: work
\&  env_parallel \-\-env multiline multiline ::: work
\&  env_parallel \-\-env multiline \-S server multiline ::: work
.Ve
.IP "functions" 8
.IX Item "functions"
.Vb 1
\&  ash cannot list defined functions \- thus is not supported.
.Ve
.IP "variables" 8
.IX Item "variables"
.Vb 5
\&  myvar=variables
\&  env_parallel echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-S server echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-\-env myvar echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-\-env myvar \-S server echo \*(Aq$myvar\*(Aq ::: work
.Ve
.IP "arrays" 8
.IX Item "arrays"
Arrays are not supported by Ash.
.SS "Bash"
.IX Subsection "Bash"
\fIInstallation\fR
.IX Subsection "Installation"
.PP
Put this in \f(CW$HOME\fR/.bashrc:
.PP
.Vb 1
\&  . \`which env_parallel.bash\`
.Ve
.PP
E.g. by doing:
.PP
.Vb 1
\&  echo \*(Aq. \`which env_parallel.bash\`\*(Aq >> $HOME/.bashrc
.Ve
.PP
\fISupported use\fR
.IX Subsection "Supported use"
.PP
\&\fB\-\-env\fR is supported to export only the variable, alias, function, or
array with the given name. Multiple \fB\-\-env\fRs can be given.
.IP "aliases" 8
.IX Item "aliases"
.Vb 5
\&  alias myecho=\*(Aqecho aliases\*(Aq
\&  env_parallel myecho ::: work
\&  env_parallel \-S server myecho ::: work
\&  env_parallel \-\-env myecho myecho ::: work
\&  env_parallel \-\-env myecho \-S server myecho ::: work
\&
\&  alias multiline=\*(Aqecho multiline
\&    echo aliases\*(Aq
\&  env_parallel \*(Aqmultiline {};
\&    echo but only when followed by a newline\*(Aq ::: work
\&  env_parallel \-S server \*(Aqmultiline {};
\&    echo but only when followed by a newline\*(Aq ::: work
\&  env_parallel \-\-env multiline \*(Aqmultiline {};
\&    echo but only when followed by a newline\*(Aq ::: work
\&  env_parallel \-\-env multiline \-S server \*(Aqmultiline {};
\&    echo but only when followed by a newline\*(Aq ::: work
.Ve
.IP "functions" 8
.IX Item "functions"
.Vb 5
\&  myfunc() { echo functions $*; }
\&  env_parallel myfunc ::: work
\&  env_parallel \-S server myfunc ::: work
\&  env_parallel \-\-env myfunc myfunc ::: work
\&  env_parallel \-\-env myfunc \-S server myfunc ::: work
.Ve
.IP "variables" 8
.IX Item "variables"
.Vb 5
\&  myvar=variables
\&  env_parallel echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-S server echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-\-env myvar echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-\-env myvar \-S server echo \*(Aq$myvar\*(Aq ::: work
.Ve
.IP "arrays" 8
.IX Item "arrays"
.Vb 6
\&  myarray=(arrays work, too)
\&  env_parallel \-k echo \*(Aq${myarray[{}]}\*(Aq ::: 0 1 2
\&  env_parallel \-k \-S server echo \*(Aq${myarray[{}]}\*(Aq ::: 0 1 2
\&  env_parallel \-k \-\-env myarray echo \*(Aq${myarray[{}]}\*(Aq ::: 0 1 2
\&  env_parallel \-k \-\-env myarray \-S server \e
\&    echo \*(Aq${myarray[{}]}\*(Aq ::: 0 1 2
.Ve
.PP
\fI\s-1BUGS\s0\fR
.IX Subsection "BUGS"
.PP
Due to a bug in Bash, aliases containing newlines must be followed by
a newline in the command.
.SS "csh"
.IX Subsection "csh"
\&\fBenv_parallel\fR for \fBcsh\fR breaks \fB\f(CB$PARALLEL\fB\fR, so do not use
\&\fB\f(CB$PARALLEL\fB\fR.
.PP
\fIInstallation\fR
.IX Subsection "Installation"
.PP
Put this in \f(CW$HOME\fR/.cshrc:
.PP
.Vb 1
\&  source \`which env_parallel.csh\`
.Ve
.PP
E.g. by doing:
.PP
.Vb 1
\&  echo \*(Aqsource \`which env_parallel.csh\`\*(Aq >> $HOME/.cshrc
.Ve
.PP
\fISupported use\fR

parallel-20180522/NEWS
20180522

* --tty allows for more programs accessing /dev/tty in parallel. Some
  programs require tty access without using it.

* env_parallel --session will record names in current environment in
  $PARALLEL_IGNORED_NAMES and exit. It is only used with env_parallel,
  and can work like --record-env but in a single session.

* 35 day course: Linux shell scripting and text processing for big
  data https://comquest.in/training

* How To Use Your Entire CPU In Bash With Parallel
  https://bash-prompt.net/guides/parallell-bash/

* An easy way to run parallel jobs on Stampede2
  https://blog.uta.edu/xus/2017/09/29/an-easy-way-to-run-parallel-jobs-on-stampede2/

* Get more done at the Linux command line with GNU Parallel
  https://opensource.com/article/18/5/gnu-parallel

* Save Time with Modern Search Techniques
  https://www.sans.org/summit-archives/file/summit-archive-1524582079.pdf

* GNU Parallel: haz más cosas simultáneamente en el terminal
  https://blog.desdelinux.net/gnu-parallel-haz-mas-cosas-simultaneamente-en-el-terminal/

* 「今日から使い切る」ための GNU Parallelによる並列処理入門
  https://www.slideshare.net/koji_matsuda/gnu-parallel

* Bug fixes and man page updates.


20180422

* --csv makes GNU Parallel parse the input sources as CSV. When used
  with --pipe it only passes full CSV-records.

* Time in --bar is printed as 1d02h03m04s.

* Optimization of --tee: It spawns a process less per value.

* The GNU Parallel 2018 book is now available:
  http://www.lulu.com/shop/ole-tange/gnu-parallel-2018/paperback/product-23558902.html

* Modern pentest tricks for faster, wider, greater engagement (15)
  https://conference.hitb.org/hitbsecconf2018ams/materials/D1%20COMMSEC%20-%20Thomas%20Debize%20-%20Modern%20Pentest%20Tricks%20for%20Faster,%20Wider,%20Greater%20Engagements.pdf

* Running many serial jobs efficiently
  https://userinfo.surfsara.nl/documentation/running-many-serial-jobs-efficiently

* GNU Parallel: как сохранить результаты нескольких команд для
  переменной? http://fliplinux.com/gnu-parallel-6.html

* Running Multiple Commands in Parallel on a GNU Linux
  https://www.youtube.com/watch?v=sHpTywpb4_4

* Klaatu covers the documentation of GNU parallel in episode 12x15
  http://gnuworldorder.info/

* Bug fixes and man page updates.


20180322

* niceload -p can now take multiple pids separated by comma

* --timeout gives a warning when killing processes

* --embed now uses the same code for all supported shells

* --delay can now take arguments like 1h12m07s

* Parallel. Straight from your command line
  https://medium.com/@alonisser/parallel-straight-from-your-command-line-feb6db8b6cee

* GNU Parallel was used in POAP-Parallelized Open Babel & Autodock
  suite Pipeline https://github.com/inpacdb/POAP

* Pi Cluster Parallel Script First Fire
  https://chiefio.wordpress.com/2018/02/23/pi-cluster-parallel-script-first-fire/

* How to Restore MySQL Logical Backup at Maximum Speed
  https://www.percona.com/blog/2018/02/22/restore-mysql-logical-backup-maximum-speed/

* Bug fixes and man page updates.


20180222

* --embed makes it possible to embed GNU parallel in a shell
  script. This is useful if you need to distribute your script to
  someone who does not want to install GNU parallel.

* Parallelize Your Large Batches of Small Jobs
  https://amito.me/2018/02/Parallelize-Your-Large-Batches-of-Small-Jobs/

* Klaatu expounds upon xargs and talks GNU parallel in episode 12x6
  http://gnuworldorder.info/

* How to use parallel to speed up your work
  https://mike42.me/blog/how-to-use-parallel-to-speed-up-your-work

* Transient compute clustering with GNU Parallel and sshfs
  https://dennismungai.wordpress.com/2018/02/06/transient-compute-clustering-with-gnu-parallel-and-sshfs/
 
* Useful Examples of ffmpeg and GNU parallel on the command-line
  https://dennismungai.wordpress.com/2018/02/06/useful-examples-of-ffmpeg-and-gnu-parallel-on-the-command-line/

* Data Science at the Command Line (Chapter 8.3)
  https://www.datascienceatthecommandline.com/

* Bug fixes and man page updates.


20180122

* bash / 병렬 https://code.i-harness.com/ko/docs/bash/2a1a

* 15分钟神器gnu parallel 入门观止
  http://www.ezilin.com/2018/01/15gnu-parallel.html

* How To Run Multiple Commands In Parallel on Linux
  https://www.slashroot.in/how-run-multiple-commands-parallel-linux

* Bug fixes and man page updates.


20171222

* env_parset for ash, dash, ksh, sh, zsh

* Automatically create hostgroups if argument ends in @sshlogin

* Tips for running jobs on your VM
  https://support.ehelp.edu.au/support/solutions/articles/6000089713-tips-for-running-jobs-on-your-vm#parallel

* Node-local scheduling
  https://mogonwiki.zdv.uni-mainz.de/dokuwiki/node_local_scheduling

* GNU Parallel https://docs.computecanada.ca/wiki/GNU_Parallel

* GNU Parallel のエラー処理 https://qiita.com/inouet/items/bfc208668c86caf8ff74

* Using GNU Parallel
  https://wiki.ncsa.illinois.edu/display/ROGER/Using+GNU+Parallel

* Bug fixes and man page updates.


20171122

* Using GNU Parallel to speed up Google Cloud Infrastructure
  management
  https://medium.com/@pczarkowski/using-gnu-parallel-to-speed-up-google-cloud-infrastructure-management-53e5c555ec05

* What is gnu parallel
  http://wiki.chpc.ac.za/howto:bioinformatics:gnu-parallel

* Parallel, Scripts, Clusters, & Easy Use
  https://chiefio.wordpress.com/2017/11/10/parallel-scripts-clusters-easy-use/

* Bug fixes and man page updates.


20171022

* Initial support for XDG Base Directory $XDG_* environment variables.

* Validating half a million TIFF files. Part
  Two
  https://www.dpoc.ac.uk/2017/08/17/validating-half-a-million-tiff-files-part-two/

* Turing data science class: Cluster computing

parallel-20180522/src/env_parallel.sh
#!/usr/bin/env sh

# This file must be sourced in sh:
#
#   . `which env_parallel.sh`
#
# after which 'env_parallel' works
#
#
# Copyright (C) 2016,2017,2018
# Ole Tange and Free Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>
# or write to the Free Software Foundation, Inc., 51 Franklin St,
# Fifth Floor, Boston, MA 02110-1301 USA

env_parallel() {
    # env_parallel.sh

    _names_of_ALIASES() {
	# alias fails on Unixware 5
	for _i in `alias 2>/dev/null | perl -ne 's/^alias //;s/^(\S+)=.*/$1/ && print' 2>/dev/null`; do
	    # Check if this name really is an alias
	    # or just part of a multiline alias definition
	    if alias $_i >/dev/null 2>/dev/null; then
		echo $_i
	    fi
	done
    }
    _bodies_of_ALIASES() {
	# alias may return:
	#   myalias='definition' (GNU/Linux ash)
	#   alias myalias='definition' (FreeBSD ash)
	# so remove 'alias ' from first line
	for _i in "$@"; do
		echo 'alias '"`alias $_i | perl -pe '1..1 and s/^alias //'`"
	done
    }
    _names_of_maybe_FUNCTIONS() {
	set | perl -ne '/^([A-Z_0-9]+)\s*\(\)\s*\{?$/i and print "$1\n"'
    }
    _names_of_FUNCTIONS() {
	# myfunc is a function
	LANG=C type `_names_of_maybe_FUNCTIONS` |
	    perl -ne '/^(\S+) is a function$/ and not $seen{$1}++ and print "$1\n"'
    }
    _bodies_of_FUNCTIONS() {
	LANG=C type "$@" | perl -ne '/^(\S+) is a function$/ or print'
    }
    _names_of_VARIABLES() {
	# This may screw up if variables contain \n and =
	set | perl -ne 's/^(\S+?)=.*/$1/ and print;'
    }
    _bodies_of_VARIABLES() {
	# Crappy typeset -p
	for _i in "$@"
	do
	    perl -e 'print @ARGV' "$_i="
	    eval echo \"\$$_i\" | perl -e '$/=undef; $a=<>; chop($a); print $a' |
		perl -pe 's/[\002-\011\013-\032\\\#\?\`\(\)\{\}\[\]\^\*\<\=\>\~\|\; \"\!\$\&\202-\377]/\\$&/go;'"s/'/\\\'/g; s/[\n]/'\\n'/go;";
	    echo
	done
    }
    _remove_bad_NAMES() {
	# Do not transfer vars and funcs from env_parallel
	# Some versions of grep do not support -E: Use perl
	# grep -Ev '^(_names_of_ALIASES|_bodies_of_ALIASES|_names_of_maybe_FUNCTIONS|_names_of_FUNCTIONS|_bodies_of_FUNCTIONS|_names_of_VARIABLES|_bodies_of_VARIABLES|_remove_bad_NAMES|_prefix_PARALLEL_ENV|_get_ignored_VARS|_make_grep_REGEXP|_ignore_UNDERSCORE|_alias_NAMES|_list_alias_BODIES|_function_NAMES|_list_function_BODIES|_variable_NAMES|_list_variable_VALUES|_prefix_PARALLEL_ENV|PARALLEL_TMP)$' |

	perl -ne '/^(_names_of_ALIASES|_bodies_of_ALIASES|_names_of_maybe_FUNCTIONS|_names_of_FUNCTIONS|_bodies_of_FUNCTIONS|_names_of_VARIABLES|_bodies_of_VARIABLES|_remove_bad_NAMES|_prefix_PARALLEL_ENV|_get_ignored_VARS|_make_grep_REGEXP|_ignore_UNDERSCORE|_alias_NAMES|_list_alias_BODIES|_function_NAMES|_list_function_BODIES|_variable_NAMES|_list_variable_VALUES|_prefix_PARALLEL_ENV|PARALLEL_TMP)$/ and next;
	    # Filter names matching --env
	    /^'"$_grep_REGEXP"'$/ or next;
            /^'"$_ignore_UNDERSCORE"'$/ and next;
	    # Vars set by /bin/sh
            /^(_|TIMEOUT)$/ and next;
            print;'
    }

    _get_ignored_VARS() {
        perl -e '
            for(@ARGV){
                $next_is_env and push @envvar, split/,/, $_;
                $next_is_env=/^--env$/;
            }
            if(grep { /^_$/ } @envvar) {
                if(not open(IN, "<", "$ENV{HOME}/.parallel/ignored_vars")) {
             	    print STDERR "parallel: Error: ",
            	    "Run \"parallel --record-env\" in a clean environment first.\n";
                } else {
            	    chomp(@ignored_vars = <IN>);
                }
            }
            if($ENV{PARALLEL_IGNORED_NAMES}) {
                push @ignored_vars, split/\s+/, $ENV{PARALLEL_IGNORED_NAMES};
                chomp @ignored_vars;
            }
            $vars = join "|",map { quotemeta $_ } "env_parallel", @ignored_vars;
	    print $vars ? "($vars)" : "(,,nO,,VaRs,,)";
            ' -- "$@"
    }

    # Get the --env variables if set
    # --env _ should be ignored
    # and convert  a b c  to (a|b|c)
    # If --env not set: Match everything (.*)
    _make_grep_REGEXP() {
        perl -e '
            for(@ARGV){
                /^_$/ and $next_is_env = 0;
                $next_is_env and push @envvar, split/,/, $_;
                $next_is_env = /^--env$/;
            }
            $vars = join "|",map { quotemeta $_ } @envvar;
            print $vars ? "($vars)" : "(.*)";
            ' -- "$@"
    }
    _which() {
	# type returns:
	#   ll is an alias for ls -l (in ash)
	#   bash is a tracked alias for /bin/bash
	#   true is a shell builtin
	#   myfunc is a function (in bash)
	#   myfunc is a shell function (in zsh)
	#   which is /usr/bin/which
	#   which is hashed (/usr/bin/which)
	#   aliased to `alias | /usr/bin/which --tty-only --read-alias --show-dot --show-tilde'
	# Return 0 if found, 1 otherwise
	LANG=C type "$@" |
	    perl -pe '$exit += (s/ is an alias for .*// ||
	                        s/ is aliased to .*// ||
                                s/ is a function// ||
                                s/ is a shell function// ||
                                s/ is a shell builtin// ||
                                s/.* is hashed .(\S+).$/$1/ ||
                                s/.* is (a tracked alias for )?//);
                      END { exit not $exit }'
    }
    _warning() {
	echo "env_parallel: Warning: $@" >&2
    }
    _error() {
	echo "env_parallel: Error: $@" >&2
    }
    
    if _which parallel >/dev/null; then
	true parallel found in path
    else
	_error 'parallel must be in $PATH.'
	return 255
    fi

    # Grep regexp for vars given by --env
    _grep_REGEXP="`_make_grep_REGEXP \"$@\"`"

    # Deal with --env _
    _ignore_UNDERSCORE="`_get_ignored_VARS \"$@\"`"

    # --record-env
    if perl -e 'exit grep { /^--record-env$/ } @ARGV' -- "$@"; then
	true skip
    else
	(_names_of_ALIASES;
	 _names_of_FUNCTIONS;
	 _names_of_VARIABLES) |
	    cat > $HOME/.parallel/ignored_vars
	return 0
    fi

    # --session
    if perl -e 'exit grep { /^--session$/ } @ARGV' -- "$@"; then
	true skip
    else
	PARALLEL_IGNORED_NAMES="`_names_of_ALIASES;
	 _names_of_FUNCTIONS;
	 _names_of_VARIABLES`"
	export PARALLEL_IGNORED_NAMES
	return 0
    fi

    # Grep alias names
    _alias_NAMES="`_names_of_ALIASES | _remove_bad_NAMES | xargs echo`"
    _list_alias_BODIES="_bodies_of_ALIASES $_alias_NAMES"
    if [ "$_alias_NAMES" = "" ] ; then
	# no aliases selected
	_list_alias_BODIES="true"
    fi
    unset _alias_NAMES

    # Grep function names
    _function_NAMES="`_names_of_FUNCTIONS | _remove_bad_NAMES | xargs echo`"
    _list_function_BODIES="_bodies_of_FUNCTIONS $_function_NAMES"
    if [ "$_function_NAMES" = "" ] ; then
	# no functions selected
	_list_function_BODIES="true"
    fi
    unset _function_NAMES

    # Grep variable names
    _variable_NAMES="`_names_of_VARIABLES | _remove_bad_NAMES | xargs echo`"
    _list_variable_VALUES="_bodies_of_VARIABLES $_variable_NAMES"
    if [ "$_variable_NAMES" = "" ] ; then
	# no variables selected
	_list_variable_VALUES="true"
    fi
    unset _variable_NAMES

    PARALLEL_ENV="`
        $_list_alias_BODIES;
        $_list_function_BODIES;
        $_list_variable_VALUES;
    `"
    export PARALLEL_ENV
    unset _list_alias_BODIES
    unset _list_variable_VALUES
    unset _list_function_BODIES
    unset _grep_REGEXP
    unset _ignore_UNDERSCORE
    # Test if environment is too big
    if `_which true` >/dev/null 2>/dev/null ; then
	parallel "$@";
	_parallel_exit_CODE=$?
	unset PARALLEL_ENV;
	return $_parallel_exit_CODE
    else
	unset PARALLEL_ENV;
	_error "Your environment is too big."
	_error "You can try 2 different approaches:"
	_error "1. Use --env and only mention the names to copy."
	_error "2. Try running this in a clean environment once:"
	_error "     env_parallel --record-env"
	_error "   And then use '--env _'"
	_error "For details see: man env_parallel"
	return 255
    fi
}

parset() {
    _parset_parallel_prg=parallel
    _parset_main "$@"
}

env_parset() {
    _parset_parallel_prg=env_parallel
    _parset_main "$@"
}

_parset_main() {
    # If $1 contains ',' or space:
    #   Split on , to get the destination variable names
    # If $1 is a single destination variable name:
    #   Treat it as the name of an array
    #
    #   # Create array named myvar
    #   parset myvar echo ::: {1..10}
    #   echo ${myvar[5]}
    #
    #   # Put output into $var_a $var_b $var_c
    #   varnames=(var_a var_b var_c)
    #   parset "${varnames[*]}" echo ::: {1..3}
    #   echo $var_c
    #
    #   # Put output into $var_a4 $var_b4 $var_c4
    #   parset "var_a4 var_b4 var_c4" echo ::: {1..3}
    #   echo $var_c4

    _parset_name="$1"
    if [ "$_parset_name" = "" ] ; then
	echo parset: Error: No destination variable given. >&2
	echo parset: Error: Try: >&2
	echo parset: Error: ' ' parset myarray echo ::: foo bar >&2
	return 255
    fi
    shift
    echo "$_parset_name" |
	perl -ne 'chomp;for (split /[, ]/) {
            # Allow: var_32 var[3]
	    if(not /^[a-zA-Z_][a-zA-Z_0-9]*(\[\d+\])?$/) {
                print STDERR "parset: Error: $_ is an invalid variable name.\n";
                print STDERR "parset: Error: Variable names must be letter followed by letters or digits.\n";
                $exitval = 255;
            }
        }
        exit $exitval;
        ' || return 255
    if perl -e 'exit not grep /,| /, @ARGV' "$_parset_name" ; then
	# $_parset_name contains , or space
	# Split on , or space to get the names
	eval "`
	    # Compute results into files
	    $_parset_parallel_prg --files -k "$@" |
		# var1= cat tmpfile1; rm tmpfile1
		# var2= cat tmpfile2; rm tmpfile2
		parallel -q echo {2}='\`cat {1}; rm {1}\`' :::: - :::+ \`
		    echo "$_parset_name" |
			perl -pe 's/,/ /g'
			 \`
	    `"
    else
	# $_parset_name does not contain , or space
	# => $_parset_name is the name of the array to put data into
	# Supported in: bash zsh ksh
	# Arrays do not work in: ash dash
	eval "$_parset_name=( $( $_parset_parallel_prg --files -k "$@" |
              perl -pe 'chop;$_="\"\`cat $_; rm $_\`\" "' ) )"
    fi
}
